### ISSUE TITLE
Array literal assignments should support Deconstruct

### ISSUE THREAD

**Issue by SeeminglyScience** [PowerShell/PowerShell#7471]
Steps to reproduce
------------------

```powershell
$tuple = [Tuple]::Create(1, 1)
$item1, $item2 = $tuple
```

Expected behavior
-----------------

`Tuple.Item1` is assigned to `$item1` and `Tuple.Item2` is assigned to `$item2`

Actual behavior
---------------

`$tuple` is assigned to `$item1`

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.4
PSEdition                      Core
GitCommitId                    6.1.0-preview.4
OS                             Microsoft Windows 10.0.15063
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

While `Tuple` types aren't used frequently in PowerShell there are a few other default types that have `Deconstruct` methods like `DictionaryEntry` and `KeyValuePair<,>`.

**Comment by BrucePay** [comment]
For tuples, i definitely agree. (In fact I think we should have more support for tuples (i.e. being able to cast an array to a tuple) being able to "index" on a tuple so $mytuple[0] is equivalent to $mytuble.Item2, etc.).

**Comment by thezim** [comment]
Would love to be able to do stuff like this.

```powershell
$domain, $username = [System.Security.Principal.WindowsIdentity]::GetCurrent().Name -split "/"
```
instead of this
```powershell
$arr = [System.Security.Principal.WindowsIdentity]::GetCurrent().Name -split "/"
$domain = $arr[0]
$username = $arr[1]
```

**Comment by iSazonov** [comment]
@thezim Your sample already works. (on Windows with -split "\\"
)

**Comment by powercode** [comment]
It would be neat to have language syntax for tuples.

**Comment by vexx32** [comment]
The title is misleading as this already works with arrays:
`$one, $two = 1,2` works as expected.

It doesn't work with tuplen, though, and I agree that should be rectified.

**Comment by SeeminglyScience** [comment]
@vexx32 Sorry for the confusion in terminology, when I refer to an array literal assignment I mean an `AssignmentStatementAst` where the left hand side is an `ArrayLiteralAst`.

**Comment by iSazonov** [comment]
Please update the issue header.

**Comment by vexx32** [comment]
An extension method? Hmm. Well, as long as it's being called from the core parsing logic side of things that should be OK since you're still in a C# realm.

Extension methods aren't available from PS itself at present, but this should still work, I think

**Comment by iSazonov** [comment]
We have an issue about extension methods.

**Comment by Jaykul** [comment]
Yeah, I don't know...

I think any feature that needs extension method support should be stuck waiting for extension method support -- like the rest of us. I don't like the idea that we might support this **one** extension method just because we wish to call it automatically 😛

**Comment by SeeminglyScience** [comment]
@Jaykul the feature request itself doesn't actually require extension method resolution, though it would make it a lot more consistent.  That said, can you elaborate a bit more on the objection?  If a system for resolving extension methods internally within the engine was added for this feature, it would partially add support with less risk. That'd make a full implementation potentially a lot easier down the road.

**Comment by SeeminglyScience** [comment]
Marking for WG review

**Comment by dkaszews** [comment]
I'm not sure we actually need extension methods, would indexers not suffice? Make `$x, $y, $z = $a` syntactically equivalent to:

```powershell
assert $a.Length -eq 3
$x = $a[0]
$y = $a[1]
$z = $a[2]
```

I find it more fitting, as it works for all ordered containers and C# `[System.Tuple]`, while also being easily implementable for custom classes without need for extension methods or `out` parameters I'm not sure powershell supports.

**Comment by dkaszews** [comment]
I think a relatively easy implementation of the above without any breaking changes would be a keyword which converts an ordered container to an `Object[]`, as those already work as expected. This can already be implemented as a simple function (phone code):

```pwsh
function unpack($obj) {
    $array = [Object[]]::new($obj.Length)
    for ($i = 0; $i -lt $obj.Length; $i += 1)  {
        $array[i] = $obj[i]
    }
    return $array 
```

Advantage of making it a keyword is that you can adjust operator precedence to avoid need of parenthesising entire expression (e.g. `= unpack Get-Foo` instead of `= unpack (Get-Foo)`), deduce type of unpacked variables for better autocomplete, and elide creating the intermediate array.

**Comment by SeeminglyScience** [comment]
Yeah I suppose in the case of tuples specifically it could just look for `ITuple` which specifies `Length` and the indexer. That shouldn't be too bad

**Comment by SeeminglyScience** [comment]
I can definitely understand the perspective but the risk is much higher. The vast majority of types will implement `IEnumerable` if it also supports those properties. If it doesn't, there may be a very good reason to avoid treating it like one. I think `Tuple` is likely the exception in this case.

**Comment by dkaszews** [comment]
@SeeminglyScience The reason why I am hesitant to go with specific interface is that I can see no common interface for all types which I consider good candidates for unpacking.

* `ITuple` has `Length` and `operator[int]`, implemented by `Tuple`
* `IEnumerable` has `Count` instead of `Length` but no indexer
* `IList` has `Count` and indexer, implemented by `List` and `Array`

There are also types such as `KeyValuePair` which implements no interface, but have member function `Deconstruct`.

Additionally, in C# ability to unpack a type is actually duck-typed, as it only checks existence of (extension) method and not an interface.

The question becomes, do we want PowerShell to be consistent with C#, or consistent with itself? Because there is already a big difference in that PowerShell allows unpacking arrays, and even throws remaining values into the last argument:

```pwsh
> $x, $y = (1, 2, 3)
> $x
1
> $y
1
2
```

whereas C# only allows unpacking types of known size and if the size does not fit, it throws a compiler error. To me, adding both `IList` and `ITuple` is the way to go.

`Deconstruct` would be a nice extra, but it requires more implementation (even `Tuple` uses extensions instead of instance methods) and discussion what to do with size mismatches, especially if there are multiple overloads.

**Comment by dkaszews** [comment]
Counterargument to my own comment about just looking at implemented properties and not interfaces: `Dictionary<int, T>` could get real weird.

**Comment by SeeminglyScience** [comment]
> @SeeminglyScience The reason why I am hesitant to go with specific interface is that I can see no common interface for all types which I consider good candidates for unpacking.

Yeah so right now it's all PowerShell's own internal logic for "is this enumerable". Which is basically `is IEnumerable and not IDictionary or string or AFewOtherThings`. Then it just enumerates and assigns accordingly

We don't want to just make tuples enumerable because that doesn't really make sense, but we still want them to work with this syntax because `$one, $two = $tuple.Item1, $tuple.Item2` is rough.

> The question becomes, do we want PowerShell to be consistent with C#, or consistent with itself?

I'd say that neither are particularly the goal of this. Consistency is nice, but rarely a driver. The main thing is "tuples and some other things that deconstruct don't currently work with our syntax, should they?"

> whereas C# only allows unpacking types of known size and if the size does not fit, it throws a compiler error. To me, adding both `IList` and `ITuple` is the way to go.
> 
> `Deconstruct` would be a nice extra, but it requires more implementation (even `Tuple` uses extensions instead of instance methods) and discussion what to do with size mismatches, especially if there are multiple overloads.

❤️ 

> Counterargument to my own comment about just looking at implemented properties and not interfaces: `Dictionary<int, T>` could get real weird.

Good catch! That is indeed a great example ❤️

**Comment by SeeminglyScience** [comment]
The Engine WG discussed this issue and concluded that we would love to see the "array assign" syntax support tuples via `ITuple`.

`Deconstruct` support would be nice but is unlikely to be an easy change does not seem to have the impact to justify its complexity. This can be revisited at a later time if sufficient impactful use cases are shown.

**Comment by dkaszews** [comment]
Bump to keep open

**Comment by RiverHeart** [comment]
> or `out` parameters I'm not sure powershell supports.

They're supported by `[ref]`. Unfortunately, script methods apparently can't be overloaded and the definition only shows as `System.Object Deconstruct();` so I guess it's ignored. Defining an overloaded `Deconstruct()` via class CodeMethod turns the signature into `static void Deconstruct(psobject value, ref item1, ref item2)` which doesn't get called automatically either.

```powershell
function New-Tuple {
    [Alias('tuple')]
    param()

    $Deconstruct = {
        param([ref] $item1, [ref] $item2)

        $item1.Value = $this.Item1
        $item2.Value = $this.Item2
        return $null
    }

    $Destructure = {
        $Foo.psobject.properties | Where-Object { $_.Name -match 'Item\d+' } | Select-Object -ExpandProperty Value
    }

    $Tuple = [System.Tuple]::Create.Invoke($args)

    Update-TypeData `
        -TypeName $Tuple.GetType().FullName `
        -MemberName 'Deconstruct' `
        -MemberType ScriptMethod `
        -Value $Deconstruct `
        -Force

    Update-TypeData `
        -TypeName $Tuple.GetType().FullName `
        -MemberName 'Destructure' `
        -MemberType ScriptMethod `
        -Value $Destructure `
        -Force

    return $Tuple
}

$Test1 = 0
$Test2 = 0
$Foo = Tuple 'one' 'two'

#$Test1, $Test2 = $foo  # Doesn't work
$Foo.Deconstruct([ref] $Test1, [ref] $Test2)  # Works
```

**Comment by mmirreza** [comment]
Since Powershell 7.4 the function [math]::DivRem() returns a tuple of the quotient and remainder. Now at the latest, we need syntax support to unpack tuples into array by assignment. Actually, it's just an operation override for assignment. Here a try with a casting type:

```
#region == C# Class Definition ==

Add-Type -TypeDefinition @"
using System;
using System.Linq;

public class NiceTuple
{
    private object[] _values;

    public NiceTuple(params object[] values)
    {
        _values = values;
    }

    public static explicit operator NiceTuple(object tuple)
    {
        if (tuple is null)
        {
            throw new ArgumentNullException(nameof(tuple), "The input object cannot be null.");
        }

        // Check if the object is a tuple using the 'is' operator
        if (!tuple.GetType().IsGenericType || tuple.GetType().GetGenericTypeDefinition() != typeof(Tuple<>).GetGenericTypeDefinition())
        {
            throw new InvalidCastException("The object is not a System.Tuple.");
        }

        // Extract tuple items dynamically
        var properties = tuple.GetType().GetProperties()
                              .Where(p => p.Name.StartsWith("Item"))
                              .OrderBy(p => p.Name)
                              .Select(p => p.GetValue(tuple))
                              .ToArray();

        return new NiceTuple(properties);
    }

    public static explicit operator object[](NiceTuple niceTuple)
    {
        return niceTuple._values;
    }

    public override string ToString()
    {
        return $"{string.Join(", ", _values)}";
    }
}
"@ -Language CSharp

#endregion == C# Class Definition ==


#region == Main ==

# Create a System.Tuple with any number of elements
$tuple = [System.Tuple]::Create(1, "Hello", 3.14, $true)

# Cast the System.Tuple to NiceTuple
$myNiceTuple = [NiceTuple]$tuple

# Unpack the NiceTuple into variables
$id, $text, $value, $flag = $myNiceTuple

#endregion == Main ==
```
